iml-gui/crate/src/components/lock_indicator.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    components::{font_awesome, popover, tooltip, Placement},
+    generated::css_classes::C,
+};
+use im::HashSet;
+use iml_wire_types::{warp_drive::Locks, LockChange, LockType, ToCompositeId};
+use seed::{prelude::*, *};
+
+pub(crate) fn lock_indicator<T>(x: &dyn ToCompositeId, locks: &Locks) -> Node<T> {
+    if let Some(lcks) = locks.get(&x.composite_id().to_string()) {
+        if !lcks.is_empty() {
+            let (rw, ro): (HashSet<&LockChange>, HashSet<&LockChange>) =
+                lcks.iter().partition(|l| l.lock_type == LockType::Write);
+
+            let mut tooltip = vec![];
+            let mut popup = vec![];
+
+            let rw_c = rw.len();
+            if rw_c > 0 {
+                if rw_c > 1 {
+                    tooltip.push(format!("{} write locks", rw_c));
+                } else {
+                    tooltip.push("1 write lock".to_string());
+                }
+                popup.push(popover::title_view("Write locks"));
+                popup.push(popover::content_view(mk_lock_list(&rw)));
+            }
+
+            let ro_c = ro.len();
+            if ro_c > 0 {
+                if ro_c > 1 {
+                    tooltip.push(format!("{} read locks", ro_c));
+                } else {
+                    tooltip.push("1 read lock".to_string());
+                }
+                popup.push(popover::title_view("Read locks"));
+                popup.push(popover::content_view(mk_lock_list(&ro)));
+            }
+
+            return span![
+                class![C.group, C.cursor_pointer, C.relative],
+                attrs! {At::TabIndex => 0},
```

Comment:
perhaps roll this and the above class block into a single fn that returns these in popover.rs. That way we won't need to repeat them each time. ref:

https://github.com/whamcloud/integrated-manager-for-lustre/blob/1b14eced92c102fddc333232706eb78f14964c97/iml-gui/crate/src/components/tooltip.rs#L7-L12

Reply:
The classes can be required for both a tooltip and a popup. I felt that it would be confusing and fragile to have only `tooltip::container()` or only `popup::container()`, or both, when the component have a popup and a tooltip, like in case of the lock indicator. I don't know what will be the best, at the moment I put it explicitly here.

Reply:
So, how about extracting to a common location and using for both. something like `group_container`.

Reply:
Next thing => https://github.com/whamcloud/integrated-manager-for-lustre/issues/1558